### PR TITLE
Fix parallelism of philosophers to 8

### PR DESCRIPTION
**Pull request by farquet**

### CODE REVIEW

**Comment by vhotspur** [comment]
Wasn't the existing setup using `Runtime.getRuntime().availableProcessors()` more flexible wrt. to different machines?

File: benchmarks/scala-stm/src/main/scala/org/renaissance/scala/stm/Philosophers.scala [hunk 1]
```diff
@@ -17,7 +17,7 @@ class Philosophers extends RenaissanceBenchmark {
   // TODO: Consolidate benchmark parameters across the suite.
   //  See: https://github.com/renaissance-benchmarks/renaissance/issues/27
 
-  private val THREAD_COUNT = Runtime.getRuntime.availableProcessors
+  private val THREAD_COUNT = 8
```
Comment (axel22) [inline benchmarks/scala-stm/src/main/scala/org/renaissance/scala/stm/Philosophers.scala]: Maybe the default should be `math.min(8, Runtime.getRuntime.availableProcessors`?

**Comment by farquet** [comment]
Yes, it's more flexible, but on big machines, it's also much more unstable, as one can see here (36 cores, 72 threads) : 
```
====== philosophers (scala-stm), iteration 0 started ======
Camera thread performed 601 scans.
====== philosophers (scala-stm), iteration 0 completed (16029.823 ms) ======
====== philosophers (scala-stm), iteration 1 started ======
Camera thread performed 384 scans.
====== philosophers (scala-stm), iteration 1 completed (9641.844 ms) ======
====== philosophers (scala-stm), iteration 2 started ======
Camera thread performed 396 scans.
====== philosophers (scala-stm), iteration 2 completed (11258.115 ms) ======
====== philosophers (scala-stm), iteration 3 started ======
Camera thread performed 457 scans.
====== philosophers (scala-stm), iteration 3 completed (12339.149 ms) ======
====== philosophers (scala-stm), iteration 4 started ======
Camera thread performed 401 scans.
====== philosophers (scala-stm), iteration 4 completed (11448.237 ms) ======
====== philosophers (scala-stm), iteration 5 started ======
Camera thread performed 388 scans.
====== philosophers (scala-stm), iteration 5 completed (11433.385 ms) ======
====== philosophers (scala-stm), iteration 6 started ======
Camera thread performed 504 scans.
====== philosophers (scala-stm), iteration 6 completed (13052.026 ms) ======
====== philosophers (scala-stm), iteration 7 started ======
Camera thread performed 446 scans.
====== philosophers (scala-stm), iteration 7 completed (12151.295 ms) ======
====== philosophers (scala-stm), iteration 8 started ======
Camera thread performed 394 scans.
====== philosophers (scala-stm), iteration 8 completed (11499.051 ms) ======
====== philosophers (scala-stm), iteration 9 started ======
Camera thread performed 479 scans.
====== philosophers (scala-stm), iteration 9 completed (13136.317 ms) ======
====== philosophers (scala-stm), iteration 10 started ======
Camera thread performed 420 scans.
====== philosophers (scala-stm), iteration 10 completed (11525.533 ms) ======
====== philosophers (scala-stm), iteration 11 started ======
Camera thread performed 396 scans.
====== philosophers (scala-stm), iteration 11 completed (11443.439 ms) ======
====== philosophers (scala-stm), iteration 12 started ======
Camera thread performed 450 scans.
====== philosophers (scala-stm), iteration 12 completed (12411.488 ms) ======
====== philosophers (scala-stm), iteration 13 started ======
Camera thread performed 402 scans.
====== philosophers (scala-stm), iteration 13 completed (11455.331 ms) ======
====== philosophers (scala-stm), iteration 14 started ======
Camera thread performed 407 scans.
====== philosophers (scala-stm), iteration 14 completed (11173.944 ms) ======
====== philosophers (scala-stm), iteration 15 started ======
Camera thread performed 402 scans.
====== philosophers (scala-stm), iteration 15 completed (11517.496 ms) ======
====== philosophers (scala-stm), iteration 16 started ======
Camera thread performed 467 scans.
====== philosophers (scala-stm), iteration 16 completed (12634.649 ms) ======
====== philosophers (scala-stm), iteration 17 started ======
Camera thread performed 495 scans.
====== philosophers (scala-stm), iteration 17 completed (13230.039 ms) ======
====== philosophers (scala-stm), iteration 18 started ======
Camera thread performed 505 scans.
====== philosophers (scala-stm), iteration 18 completed (13469.568 ms) ======
====== philosophers (scala-stm), iteration 19 started ======
Camera thread performed 432 scans.
====== philosophers (scala-stm), iteration 19 completed (12031.045 ms) ======
====== philosophers (scala-stm), iteration 20 started ======
Camera thread performed 430 scans.
====== philosophers (scala-stm), iteration 20 completed (11963.825 ms) ======
====== philosophers (scala-stm), iteration 21 started ======
Camera thread performed 563 scans.
====== philosophers (scala-stm), iteration 21 completed (14600.265 ms) ======
====== philosophers (scala-stm), iteration 22 started ======
Camera thread performed 574 scans.
====== philosophers (scala-stm), iteration 22 completed (14516.369 ms) ======
====== philosophers (scala-stm), iteration 23 started ======
Camera thread performed 425 scans.
====== philosophers (scala-stm), iteration 23 completed (11311.409 ms) ======
====== philosophers (scala-stm), iteration 24 started ======
Camera thread performed 452 scans.
====== philosophers (scala-stm), iteration 24 completed (13003.634 ms) ======
====== philosophers (scala-stm), iteration 25 started ======
Camera thread performed 416 scans.
====== philosophers (scala-stm), iteration 25 completed (11572.823 ms) ======
====== philosophers (scala-stm), iteration 26 started ======
Camera thread performed 549 scans.
====== philosophers (scala-stm), iteration 26 completed (14010.356 ms) ======
====== philosophers (scala-stm), iteration 27 started ======
Camera thread performed 547 scans.
====== philosophers (scala-stm), iteration 27 completed (14211.133 ms) ======
====== philosophers (scala-stm), iteration 28 started ======
Camera thread performed 461 scans.
====== philosophers (scala-stm), iteration 28 completed (12120.816 ms) ======
====== philosophers (scala-stm), final iteration started ======
Camera thread performed 402 scans.
====== philosophers (scala-stm), final iteration completed (11008.325 ms) ======
```
Fixing to 8 gives : 
```
...
====== philosophers (scala-stm), iteration 22 started ======
Camera thread performed 243 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 22 completed (4320.565 ms) ======
====== philosophers (scala-stm), iteration 23 started ======
Camera thread performed 255 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 23 completed (4189.521 ms) ======
====== philosophers (scala-stm), iteration 24 started ======
Camera thread performed 270 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 24 completed (4642.774 ms) ======
====== philosophers (scala-stm), iteration 25 started ======
Camera thread performed 252 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 25 completed (4320.488 ms) ======
====== philosophers (scala-stm), iteration 26 started ======
Camera thread performed 243 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 26 completed (3985.315 ms) ======
====== philosophers (scala-stm), iteration 27 started ======
Camera thread performed 219 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 27 completed (3913.676 ms) ======
====== philosophers (scala-stm), iteration 28 started ======
Camera thread performed 266 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 28 completed (4648.248 ms) ======
====== philosophers (scala-stm), final iteration started ======
Camera thread performed 242 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), final iteration completed (4242.179 ms) ======
```

File: benchmarks/scala-stm/src/main/scala/org/renaissance/scala/stm/Philosophers.scala (hunk 1 above)
Comment (farquet) [reply benchmarks/scala-stm/src/main/scala/org/renaissance/scala/stm/Philosophers.scala]: Why not. I can change this.
I like this strategy for the rework that will be necessary for #27 .

**Comment by vhotspur** [comment]
>  Yes, it's more flexible, but on big machines, it's also much more unstable, as one can see here (36 cores, 72 threads) 

Isn't that an inherent property of the benchmark (as the threads do not yield)? Computing relative variability of the data, it goes from 5% to 10% (and there are almost 10 times more threads in the second case). So I would say it is not that bad.

Furthermore my guess is that this is more affected by OS scheduler rather than anything in JVM so why not to keep it as parallel as possible?

**Comment by axel22** [comment]
> > Yes, it's more flexible, but on big machines, it's also much more unstable, as one can see here (36 cores, 72 threads)
> 
> Isn't that an inherent property of the benchmark (as the threads do not yield)? Computing relative variability of the data, it goes from 5% to 10% (and there are almost 10 times more threads in the second case). So I would say it is not that bad.
> 
> Furthermore my guess is that this is more affected by OS scheduler rather than anything in JVM so why not to keep it as parallel as possible?

It does seem that the occasional jump to 14s in the 72-thread version is perhaps more GC-related, rather than contention-related. Perhaps this benchmark should be kept at the max, but just have its default iteration count increased?

**Comment by axel22** [comment]
Or `-verbose:gc -XX:+PrintGCDetails`.

**Comment by farquet** [comment]
GC log of all iterations look exactly the same : 
```
====== philosophers (scala-stm), iteration 27 completed (16431.66 ms) ======
====== philosophers (scala-stm), iteration 28 started ======
[GC (Allocation Failure) [PSYoungGen: 2779552K->3136K(2782208K)] 2795802K->19386K(8374784K), 0.0119638 secs] [Times: user=0.29 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2771520K->3680K(2782208K)] 2787770K->19930K(8374784K), 0.0101530 secs] [Times: user=0.23 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2772064K->4480K(2782208K)] 2788314K->20738K(8374784K), 0.0095391 secs] [Times: user=0.27 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2772864K->5184K(2782208K)] 2789122K->21442K(8374784K), 0.0110564 secs] [Times: user=0.27 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2773568K->5728K(2783232K)] 2789826K->21986K(8375808K), 0.0088856 secs] [Times: user=0.22 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2775648K->6592K(2782720K)] 2791906K->22850K(8375296K), 0.0104848 secs] [Times: user=0.26 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2776512K->7072K(2784256K)] 2792770K->23330K(8376832K), 0.0088177 secs] [Times: user=0.21 sys=0.02, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2779040K->7808K(2783744K)] 2795298K->24066K(8376320K), 0.0095422 secs] [Times: user=0.22 sys=0.02, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2779776K->8544K(2783232K)] 2796034K->24802K(8375808K), 0.0114255 secs] [Times: user=0.30 sys=0.02, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2780000K->9024K(2783744K)] 2796258K->25282K(8376320K), 0.0103759 secs] [Times: user=0.27 sys=0.00, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2780480K->9536K(2782720K)] 2796738K->25794K(8375296K), 0.0121788 secs] [Times: user=0.27 sys=0.01, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2779968K->10208K(2780672K)] 2796226K->26466K(8373248K), 0.0108516 secs] [Times: user=0.21 sys=0.02, real=0.01 secs]
[GC (Allocation Failure) [PSYoungGen: 2780640K->10048K(2781696K)] 2796898K->26306K(8374272K), 0.0105093 secs] [Times: user=0.26 sys=0.00, real=0.01 secs]
Camera thread performed 864 scans.
WARNING: This benchmark provides no result that can be validated.
         There is no way to check that no silent failure occurred.
====== philosophers (scala-stm), iteration 28 completed (17286.568 ms) ======
```
Looks like we are allocating a lot, but a lot can be thrown away too.

More iterations doesn't give more stability from the experiments I made. Alternatively, we could reduce the number of meals we serve to the philosophers ? 500000 seems quite generous.

**Comment by axel22** [comment]
So there are no "Full GC" log outputs?

**Comment by farquet** [comment]
> So there are no "Full GC" log outputs?
Correct, none throughout the whole execution.

**Comment by vhotspur** [comment]
> > So there are no "Full GC" log outputs?
> 
> Correct, none throughout the whole execution.

Looking at the sources, it seems that some allocations come from the use of `Some` wrapper for the fork owner (but replacing it with plain Java `null` does not change the GC output much). So the majority is probably coming from inside STM as each `atomic {}` block with `retry` allocates a lot of state objects when doing the rollback. And as this benchmark does nothing else, there will be plenty of them. Not much we can do about this IMHO.

**Comment by farquet** [comment]
> Still I believe that if the benchmark is supposed to demonstrate how the library/JVM behaves under high load (with highly competitive threads), it should run on all cores.

Agreed. So I'll close the PR accepting that this benchmark is a bit more unstable than others.

**Pull request closed without merging**